File: ui/js/devices.js


// Call the dataTables jQuery plugin
function getIdSelections(table) {
  return $.map(table.bootstrapTable('getSelections'), function (row) {
    return row.internal_ip
  })
}

function responseHandler(res) {
  $.each(res.rows, function (i, row) {
    row.state = $.inArray(row.internal_ip, selections) !== -1
  })
  return res
}


$(function () {
  let table = createTable('#devicesTable', [
    {
      field: 'state',
      checkbox: true,
      align: 'center',
    }, {
      title: 'Owner',
      field: 'owner',
      align: 'center',
      sortable: true,
    }, {
      field: 'is_locked',
      title: 'Locked',
      sortable: true,
      align: 'center'
    }, {
      field: 'internal_ip',
      title: 'Address',
      sortable: true,
      align: 'center'
    }, {
      field: 'public_key',
      title: 'Public Key',
      sortable: true,
      align: 'center'
    }, {
      field: 'last_endpoint',
      title: 'Last Endpoint Address',
      sortable: true,
      align: 'center'
    }
  ])

  var $remove = $('#remove')
  var $lock = $('#lock')
  var $unlock = $('#unlock')


  table.on('check.bs.table uncheck.bs.table ' +
    'check-all.bs.table uncheck-all.bs.table',
    function () {
      let enableModifications = !table.bootstrapTable('getSelections').length;

      $("#removeStart").prop('disabled', enableModifications)
      $lock.prop('disabled', enableModifications)
      $unlock.prop('disabled', enableModifications)

      // save your data, here just save the current page
      selections = getIdSelections(table)
      // push or splice the selections if you want to save all data selections
    })
  $lock.on("click", function () {
    var ids = getIdSelections(table)
    action(ids, "lock", table)
  })

  $unlock.on("click", function () {
    var ids = getIdSelections(table)
    action(ids, "unlock", table)
  })

  $remove.on("click", function () {
    var ids = getIdSelections(table)
    table.bootstrapTable('remove', {
      field: 'internal_ip',
      values: ids
    })


    fetch("/management/devices/data", {
      method: 'DELETE',
      mode: 'same-origin',
      cache: 'no-cache',
      credentials: 'same-origin',
      redirect: 'follow',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(ids)
    }).then((response) => {
      if (response.status == 200) {
        table.bootstrapTable('refresh')
        $("#issue").hide()
        return
      }

      response.text().then(txt => {

        $("#issue").text(txt)
        $("#issue").show()
      })
    })
  })

});

function action(onDevices, action, table) {
  let data = {
    "action": action,
    "addresses": onDevices,
  }

  fetch("/management/devices/data", {
    method: 'PUT',
    mode: 'same-origin',
    cache: 'no-cache',
    credentials: 'same-origin',
    redirect: 'follow',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(data)
  }).then((response) => {
    if (response.status == 200) {
      table.bootstrapTable('refresh')
      $("#issue").hide()
      return
    }

    response.text().then(txt => {
      $("#issue").text(txt)
      $("#issue").show()
    })
  })
}

File: ui/devices.go
package ui

import (
	"encoding/json"
	"log"
	"net/http"
)

func devicesMgmtUI(w http.ResponseWriter, r *http.Request) {

	_, u := sessionManager.GetSessionFromRequest(r)
	if u == nil {
		http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
		return
	}

	d := Page{

		Description:  "Devices Management Page",
		Title:        "Devices",
		User:         u.Username,
		WagVersion:   WagVersion,
		ServerID:     serverID,
		ClusterState: clusterState,
	}

	err := renderDefaults(w, r, d, "management/devices.html", "delete_modal.html")

	if err != nil {
		log.Println("unable to render devices page: ", err)

		w.WriteHeader(http.StatusInternalServerError)
		renderDefaults(w, r, nil, "error.html")
		return
	}
}

func devicesMgmt(w http.ResponseWriter, r *http.Request) {

	switch r.Method {
	case "GET":
		allDevices, err := ctrl.ListDevice("")
		if err != nil {
			log.Println("error getting devices: ", err)

			w.WriteHeader(http.StatusInternalServerError)
			renderDefaults(w, r, nil, "error.html")
			return
		}

		lockout, err := ctrl.GetLockout()
		if err != nil {
			log.Println("error getting lockout: ", err)

			w.WriteHeader(http.StatusInternalServerError)
			renderDefaults(w, r, nil, "error.html")
			return
		}

		var deviceData []DevicesData

		for _, dev := range allDevices {
			deviceData = append(deviceData, DevicesData{
				Owner:        dev.Username,
				Locked:       dev.Attempts >= lockout,
				InternalIP:   dev.Address,
				PublicKey:    dev.Publickey,
				LastEndpoint: dev.Endpoint.String(),
				Active:       dev.Active,
			})
		}

		b, err := json.Marshal(deviceData)
		if err != nil {

			log.Println("unable to marshal devices data: ", err)
			http.Error(w, "Server error", 500)

			return
		}

		w.Header().Set("Content-Type", "application/json")
		w.Write(b)
	case "PUT":
		var action struct {
			Action    string   `json:"action"`
			Addresses []string `json:"addresses"`
		}

		err := json.NewDecoder(r.Body).Decode(&action)
		if err != nil {
			http.Error(w, "Bad request", 400)
			return
		}

		for _, address := range action.Addresses {
			switch action.Action {
			case "lock":
				err := ctrl.LockDevice(address)
				if err != nil {
					log.Println("Error locking device: ", address, " err:", err)
					http.Error(w, err.Error(), http.StatusInternalServerError)
					return
				}
			case "unlock":
				err := ctrl.UnlockDevice(address)
				if err != nil {
					log.Println("Error unlocking device: ", address, " err:", err)
					http.Error(w, err.Error(), http.StatusInternalServerError)
					return
				}
			default:
				http.Error(w, "invalid action", 400)
				return
			}
		}

		w.Write([]byte("OK"))

	case "DELETE":
		var addresses []string

		err := json.NewDecoder(r.Body).Decode(&addresses)
		if err != nil {
			http.Error(w, "Bad request", 400)
			return
		}

		for _, address := range addresses {
			err := ctrl.DeleteDevice(address)
			if err != nil {
				log.Println("Error Deleting device: ", address, "err:", err)

				http.Error(w, err.Error(), http.StatusInternalServerError)
				return
			}
		}
		w.Write([]byte("OK"))

	default:
		http.NotFound(w, r)
	}

}
